### ISSUE TITLE
Enhancement/Question on event loop time

### ISSUE THREAD

**Issue by pioardi** [netty/netty#12305]
Hi and thanks to create this amazing piece of software, I was wondering if is possible ( or if it is already in place ) to have a mechanism that logs a WARN message when an event loop cycle takes too much time and/or log the cycle time of the event loop.
This would help people using Netty to figure out when the event loop is used not appropriately, the cycle time may be also a metrics to push somewhere.

I am open to contribute on that but could need some guidance on the implementation.

Thanks in advance

**Comment by franz1981** [comment]
Hi @pioardi you can just provide your Runnable abstraction that could (with sampling; meaning that not every Runnable perform this metric) measure the time spent while it's executed.
It won't give you how much time is spent while executing application tasks, but it would save Netty to provide such functionality that can have a performance impact.

There's a GSoC in the past aiming to provide lightweight metrics; if you're interested maybe we can think about resurrecting it...
and we *love* PRs if you wish to contribute on it :P

**Comment by pioardi** [comment]
Hei @franz1981  thanks for your quick reply, yes I think lightweight metrics would be nice to have with a possibility to enable and disable them so that who is interested can turn them on, do you have any starting point where I can start looking into ?

**Comment by franz1981** [comment]
Yes @pioardi 

- https://github.com/netty/netty/wiki/Google-Summer-of-Code-Ideas-2020#implement-event-loop-telemetry
- https://github.com/netty/netty/pull/11293
- https://github.com/netty/netty/pull/9080

And IIRC https://github.com/eclipse-vertx/vert.x/blob/d711e50174f84a06744c855dd29222beaa97bdeb/src/main/java/io/vertx/core/impl/btc/BlockedThreadChecker.java VErt-x (nudge nudge @vietj ) is using this mechanism

I can help to implement something, wdyt @chrisvest @normanmaurer @NiteshKant  ?

**Comment by pioardi** [comment]
Yep the vertx implementation is exactly what I have in mind, we may start small from that and then add the metrics later ?

**Comment by franz1981** [comment]
Re 

> that can be captured only within Netty

I mean, for example, to expose hooks to report the time spent while performing I/O vs application code eg https://github.com/netty/netty/blob/86004b7303621784f7ec8850323d9a707f17f9f8/transport-classes-epoll/src/main/java/io/netty/channel/epoll/EpollEventLoop.java#L383-L391

These are not meant to be checks that can alert anyone and will be published when they happen and complete (because captured within the event loop); it's not meant to poll if an existing task is still blocking the event loop while it's happening, as the Vert-x checker.

**Comment by franz1981** [comment]
@violetagg too, anything that can be interesting for you as well?
